src/utils/mergeOptions.ts

```diff
@@ -38,160 +96,164 @@ export default function mergeOptions({
-				? Object.assign({}, configOption, commandOption)
-				: commandOption;
-		}
-		return configOption;
+	const deprecations = deprecate(config, rawCommandOptions, deprecateConfig);
+	const command = getCommandOptions(rawCommandOptions);
+	const inputOptions = getInputOptions(config, command, defaultOnWarnHandler);
+
+	if (command.output) {
+		Object.assign(command, command.output);
 	}
 
-	const onwarn = config.onwarn;
-	let warn: WarningHandler;
+	const normalizedOutputOptions = ensureArray(config.output);
+	if (normalizedOutputOptions.length === 0) normalizedOutputOptions.push({});
+	const outputOptions = normalizedOutputOptions.map(singleOutputOptions =>
+		getOutputOptions(singleOutputOptions, command)
+	);
 
-	if (onwarn) {
-		warn = warning => onwarn(warning, defaultOnWarnHandler);
-	} else {
-		warn = defaultOnWarnHandler;
-	}
+	const unknownOptionErrors: string[] = [];
+	const validInputOptions = Object.keys(inputOptions);
+	addUnknownOptionErrors(
+		unknownOptionErrors,
+		Object.keys(config),
+		validInputOptions,
+		'input',
+		/^output$/
+	);
 
-	const inputOptions: InputOptions = {
-		acorn: config.acorn,
-		acornInjectPlugins: config.acornInjectPlugins,
-		cache: getInputOption('cache'),
-		context: config.context,
-		experimentalCodeSplitting: getInputOption('experimentalCodeSplitting'),
-		experimentalDynamicImport: getInputOption('experimentalDynamicImport'),
-		experimentalPreserveModules: getInputOption('experimentalPreserveModules'),
-		input: getInputOption('input'),
-		legacy: getInputOption('legacy'),
-		moduleContext: config.moduleContext,
-		onwarn: warn,
-		perf: getInputOption('perf'),
-		plugins: config.plugins,
-		preferConst: getInputOption('preferConst'),
-		preserveSymlinks: getInputOption('preserveSymlinks'),
-		treeshake: getObjectOption('treeshake'),
-		watch: config.watch
+	const validOutputOptions = Object.keys(outputOptions[0]);
+	addUnknownOptionErrors(
+		unknownOptionErrors,
+		outputOptions.reduce((allKeys, options) => allKeys.concat(Object.keys(options)), []),
+		validOutputOptions,
+		'output'
+	);
+
+	addUnknownOptionErrors(
+		unknownOptionErrors,
+		Object.keys(command),
+		validInputOptions.concat(
+			validOutputOptions,
+			Object.keys(commandAliases),
+			'config',
+			'environment',
+			'silent'
+		),
+		'CLI',
+		/^_|output|(config.*)$/
+	);
+
+	return {
+		inputOptions,
+		outputOptions,
+		deprecations,
+		optionError: unknownOptionErrors.length > 0 ? unknownOptionErrors.join('\n') : null
 	};
+}
 
-	// legacy, to ensure e.g. commonjs plugin still works
-	(<any>inputOptions).entry = inputOptions.input;
+function addUnknownOptionErrors(
+	errors: string[],
+	options: string[],
+	validOptions: string[],
+	optionType: string,
+	ignoredKeys: RegExp = /$./
+) {
+	const unknownOptions = options.filter(
+		key => validOptions.indexOf(key) === -1 && !ignoredKeys.test(key)
+	);
+	if (unknownOptions.length > 0)
+		errors.push(
+			`Unknown ${optionType} option: ${unknownOptions.join(
```

Comment:
Very minor point with the wording here for `Unknown CLI option:`... perhaps make this `Unknown ${optionType}:` where `optionType` is `'CLI flag'` or `'input option'` or `'output option'`.

Reply:
👍